File: plugins/approvej-intellij-plugin/src/main/java/org/approvej/intellij/ApproveCallUtil.java
package org.approvej.intellij;

import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiMethod;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.uast.UCallExpression;
import org.jetbrains.uast.UElement;
import org.jetbrains.uast.UExpression;
import org.jetbrains.uast.UQualifiedReferenceExpression;
import org.jetbrains.uast.UastUtils;

/** Shared utilities for recognising and analysing {@code approve()} call chains. */
final class ApproveCallUtil {

  static final String APPROVAL_BUILDER_CLASS = "org.approvej.ApprovalBuilder";

  private ApproveCallUtil() {}

  /**
   * Returns the UAST call expression if the given PSI element is the identifier of an {@code
   * approve()} call on {@code ApprovalBuilder}, or {@code null} otherwise.
   */
  static @Nullable UCallExpression asApproveCall(@NotNull PsiElement element) {
    if (element.getFirstChild() != null) return null;

    UElement uElement = UastUtils.getUParentForIdentifier(element);
    if (!(uElement instanceof UCallExpression callExpression)) return null;
    return isApproveCall(callExpression) ? callExpression : null;
  }

  /**
   * Returns {@code true} if the given {@code approve()} call is on {@code ApprovalBuilder}.
   *
   * <p>Unlike {@link #asApproveCall}, this takes a UAST node directly (no leaf-element check).
   */
  static boolean isApproveCall(@NotNull UCallExpression node) {
    if (!"approve".equals(node.getMethodName())) return false;

    PsiMethod method = node.resolve();
    if (method == null) return false;
    PsiClass containingClass = method.getContainingClass();
    return containingClass != null
        && APPROVAL_BUILDER_CLASS.equals(containingClass.getQualifiedName());
  }

  /**
   * Walks the UAST parent chain from an {@code approve()} call and returns the name of the last
   * method in the chain (e.g. {@code "byFile"}, {@code "printedAs"}, or {@code "approve"} if there
   * is no chained call).
   */
  static ChainWalkResult findTerminalCall(@NotNull UCallExpression approveCall) {
    UElement current = approveCall;
    String lastMethodName = "approve";
    while (true) {
      UElement parent = current.getUastParent();
      if (parent instanceof UQualifiedReferenceExpression qualRef) {
        UExpression selector = qualRef.getSelector();
        if (selector instanceof UCallExpression selectorCall
            && selectorCall.getMethodName() != null) {
          lastMethodName = selectorCall.getMethodName();
          current = qualRef;
          continue;
        }
      }
      break;
    }
    return new ChainWalkResult(lastMethodName, current);
  }

  /**
   * Walks the UAST parent chain from an {@code approve()} call and returns the string literal
   * argument of a {@code .named()} call in the chain, or {@code null} if no such call exists.
   */
  static @Nullable String findNamedArgument(@NotNull UCallExpression approveCall) {
    UElement current = approveCall;
    while (true) {
      UElement parent = current.getUastParent();
      if (parent instanceof UQualifiedReferenceExpression qualRef) {
        UExpression selector = qualRef.getSelector();
        if (selector instanceof UCallExpression selectorCall
            && "named".equals(selectorCall.getMethodName())) {
          List<UExpression> args = selectorCall.getValueArguments();
          if (args.size() == 1) {
            Object evaluated = args.getFirst().evaluate();
            if (evaluated instanceof String s) {
              return s;
            }
          }
        }
        current = qualRef;
        continue;
      }
      break;
    }
    return null;
  }

  record ChainWalkResult(String lastMethodName, UElement chainEnd) {}
}

File: plugins/approvej-intellij-plugin/src/main/kotlin/org/approvej/intellij/ApproveCallUtil.kt
package org.approvej.intellij

import com.intellij.psi.PsiElement
import org.jetbrains.uast.UCallExpression
import org.jetbrains.uast.UElement
import org.jetbrains.uast.UExpression
import org.jetbrains.uast.UQualifiedReferenceExpression
import org.jetbrains.uast.getUParentForIdentifier

/** Shared utilities for recognising and analysing `approve()` call chains. */
internal object ApproveCallUtil {

  const val APPROVAL_BUILDER_CLASS = "org.approvej.ApprovalBuilder"
  const val IMAGE_APPROVAL_BUILDER_CLASS = "org.approvej.image.ImageApprovalBuilder"

  /**
   * Returns the UAST call expression if the given PSI element is the identifier of an `approve()`
   * call on `ApprovalBuilder`, or `null` otherwise.
   */
  fun asApproveCall(element: PsiElement): UCallExpression? {
    if (element.firstChild != null) return null
    val uElement = getUParentForIdentifier(element)
    return (uElement as? UCallExpression)?.takeIf { isApproveCall(it) }
  }

  /**
   * Returns `true` if the given `approve()` call is on `ApprovalBuilder`.
   *
   * Unlike [asApproveCall], this takes a UAST node directly (no leaf-element check).
   */
  fun isApproveCall(node: UCallExpression): Boolean {
    val name = node.methodName
    if (name != "approve" && name != "approveImage") return false
    val method = node.resolve() ?: return false
    val containingClass = method.containingClass ?: return false
    val qname = containingClass.qualifiedName
    return qname == APPROVAL_BUILDER_CLASS || qname == IMAGE_APPROVAL_BUILDER_CLASS
  }

  /**
   * Walks the UAST parent chain from an `approve()` call and returns the name of the last method in
   * the chain (e.g. `"byFile"`, `"printedAs"`, or `"approve"` if there is no chained call).
   */
  fun findTerminalCall(approveCall: UCallExpression): ChainWalkResult {
    var lastMethodName = approveCall.methodName ?: "approve"
    val chainEnd =
      walkChain(approveCall) { selector ->
        val name = selector.methodName
        if (name != null) {
          lastMethodName = name
          true
        } else {
          false
        }
      }
    return ChainWalkResult(lastMethodName, chainEnd)
  }

  /**
   * Walks the UAST parent chain from an `approve()` call and returns the string literal argument of
   * a `.named()` call in the chain, or `null` if no such call exists.
   */
  fun findNamedArgument(approveCall: UCallExpression): String? {
    walkChain(approveCall) { selector ->
      if (selector.methodName == "named") {
        val args: List<UExpression> = selector.valueArguments
        if (args.size == 1) {
          val evaluated = args.first().evaluate()
          if (evaluated is String) return evaluated
        }
      }
      true
    }
    return null
  }

  /**
   * Returns `true` if the UAST parent chain from an `approve()` call contains a `.named()` call,
   * regardless of whether the argument is a constant.
   */
  fun hasNamedCall(approveCall: UCallExpression): Boolean {
    walkChain(approveCall) { selector ->
      if (selector.methodName == "named") return true
      true
    }
    return false
  }

  /**
   * Walks the qualified-reference parent chain from [start], calling [visitSelector] for each
   * chained [UCallExpression]. The visitor returns `true` to continue walking or `false` to stop.
   * Returns the outermost [UElement] reached.
   */
  private inline fun walkChain(
    start: UCallExpression,
    visitSelector: (UCallExpression) -> Boolean,
  ): UElement {
    var current: UElement = start
    while (true) {
      val parent = current.uastParent
      if (parent is UQualifiedReferenceExpression) {
        val selector = parent.selector
        if (selector is UCallExpression) {
          if (!visitSelector(selector)) break
          current = parent
          continue
        }
      }
      break
    }
    return current
  }

  data class ChainWalkResult(val lastMethodName: String, val chainEnd: UElement)
}
